### PR TITLE
#117 Use installed.json for package discovery

### PR DESCRIPTION
**Pull request by jtheuerkauf**
Replace namespace source with root `composer.json` `"autoload"` data and `vendor/composer/installed.json` `"packages"` data

Root `composer.json` includes `"autoload-dev"` conditionally on `ENVIRONMENT` value

### CODE REVIEW

File: src/Application/Kernel.php
Comment (aidan-casey) [inline src/Application/Kernel.php:159]: So here, I'd be curious what something to the effect of the following would give you. 🙂 

```php
$installedContents = file_get_contents($installed);

// Splits each package section {} into it's own block
preg_match_all('/(?<=\s{8})\{[\s\S]+?\n\s{8}\}/', $installedContents, $packages);

foreach ($packages as $package) {
    // Decodes for the specific package
    $package = json_decode($package, true);
}
```

Comment (jtheuerkauf) [reply src/Application/Kernel.php:159]: i'll give it a shot and see what happens.

Comment (jtheuerkauf) [reply src/Application/Kernel.php:159]: ## Results

Also added `laravel/framework` to the Tempest `composer.json` (HOLY CRAP Symfony, are you entirely built on polyfills???)

Using the regex:
```php
if (is_file($installed)) {
    $installed = file_get_contents($installed);
    preg_match_all('~(?<=\s{8})\{[\s\S]+?\n\s{8}}~', $installed, $packages);

    foreach ($packages[0] as $package) {
        $package = json_decode($package, true);
        $packagePath = realpath(path($composerPath, $package['install-path']));
```

Benchmarks for regex and my version with Laravel included...

My version:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done

0.22264194488525
0.22401404380798
0.27412104606628
0.24248886108398
0.22104001045227
0.26659607887268
0.20401000976562
0.2078320980072
0.18999195098877
0.23425197601318

(Avg: 0.228 sec)
```

Regex:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done

0.27564597129822
0.25507497787476
0.26790499687195
0.21789503097534
0.20109605789185
0.19769716262817
0.25411415100098
0.21002793312073
0.22865700721741
0.25373101234436

(Avg: 0.236 sec)
```

Roughly speaking, very little performance change. By the way, that's ~100 total package entries in `installed.json` so i imagine it gives a better than average estimation. 😃

Comment (jtheuerkauf) [reply src/Application/Kernel.php:159]: i've seen references to more advanced JSON handlers that might help performance, but haven't worked with any of them previously, so maybe it's something we can try in a later revision. Let me know whether to update this PR to the regex or not.

Comment (aidan-casey) [reply src/Application/Kernel.php:159]: Okay, so I am getting very different numbers from you. I started the evening genuinely curious and confused and ended the evening over-optimizing microseconds. 😅 

For some reason I cannot replicate your timings at all. I don't know if it is a PHP version mismatch (I am on 8.3) or if you are running on a slower disk (HDD?). See below for the timing I am currently getting and the "optimized" code I am currently using.


```bash
# json_decode
0.0014290809631348n
0.0011060237884521n
0.0010828971862793n
0.0011448860168457n
0.0010919570922852n
0.0010881423950195n
0.0010750293731689n
0.0010759830474854n
0.0011980533599854n
0.0010321140289307n

# RegEx
0.00080204010009766n
0.000762939453125n
0.00078010559082031n
0.00077986717224121n
0.00078606605529785n
0.00073504447937012n
0.0007779598236084n
0.000762939453125n
0.00078701972961426n
0.00080013275146484n
```

```php
private function discoverInstalledPackageLocations(): void
{
    $startup = microtime(true);

    $composerPath = path($this->root, 'vendor/composer');
    $installed = path($composerPath, 'installed.json');

    if (! is_file($installed)) {
        return;
    }

    $content = file_get_contents($installed);
    $packages = json_decode($content, true)['packages'];

    foreach ($packages as $package) {
        $packagePath = realpath(path($composerPath, $package['install-path']));

        if (
            ! isset($package['require']['tempest/framework']) ||
            empty($package['autoload']['psr-4'])
        ) {
            continue;
        }

        foreach ($package['autoload']['psr-4'] as $namespace => $namespacePath) {
            $namespacePath = path($packagePath, $namespacePath);
            $this->addDiscoveryLocation($namespace, $namespacePath);
        }
    }

    die(microtime(true) - $startup);
}
```

```php
private function discoverInstalledPackageLocations(): void
{
    $startup = microtime(true);

    $composerPath = path($this->root, 'vendor/composer');
    $installed = path($composerPath, 'installed.json');

    if (! is_file($installed)) {
        return;
    }

    $content = file_get_contents($installed);

    preg_match_all('/\s{8}\{.*?tempest\/framework.*?\n\s{8}\}/s', $content, $packages);

    foreach ($packages[0] as $package) {
        $package = json_decode($package, true);
        $packagePath = realpath(path($composerPath, $package['install-path']));

        if (
            ! isset($package['require']['tempest/framework']) ||
            empty($package['autoload']['psr-4'])
        ) {
            continue;
        }

        foreach ($package['autoload']['psr-4'] as $namespace => $namespacePath) {
            $namespacePath = path($packagePath, $namespacePath);
            $this->addDiscoveryLocation($namespace, $namespacePath);
        }
    }

    die(microtime(true) - $startup);
}
```

Full disclosure, I did adjust my RegEx slightly so that now it is _only_ pulling the packages that mention Tempest in. It did not improve speed dramatically, but it is a difference from yours.

Comment (brendt) [reply src/Application/Kernel.php:159]: I suspect the difference for regex probably has to do with differences in CPU. I still don't think regex is the way to go, as we'll probably miss edge cases leading to subtle bugs.

Comment (jtheuerkauf) [reply src/Application/Kernel.php:159]: Yeah, very weird. i'm running 8.3 through WSL. i disabled all extensions to the minimum that would run the script (mbstring and pdo)...

Your version with `json_decode()`:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done

0.21244215965271
0.19042801856995
0.19336795806885
0.19525384902954
0.179358959198
0.20081305503845
0.20244216918945
0.18885803222656
0.22774887084961
0.21510100364685
```

Your version with regex:
```
$ for ((i = 0; i < 10; i++)); do php tempest; done
0.57280683517456
0.57812809944153
0.58618807792664
0.56667304039001
0.58222198486328
0.5756139755249
0.5791130065918
0.5681848526001
0.56858205795288
0.57621693611145
```
That difference from your numbers is _crazy_ ...

i'll put up results using 8.3 from PowerShell in a separate comment.

Comment (aidan-casey) [reply src/Application/Kernel.php:159]: @jtheuerkauf - Are you using WSL 2 and have the files on /mnt/c? If so, you are probably running into the slowness of the 9P protocol.

Comment (jtheuerkauf) [reply src/Application/Kernel.php:159]: Files are on a different mount, but otherwise normal. You may be right though. Here's 8.3 from PS:

`json_decode()`:
```
> for ($i = 0; $i -lt 10; $i++) { php .\tempest }

0.004680871963501
0.0046839714050293
0.0047390460968018
0.0050721168518066
0.0051999092102051
0.0046918392181396
0.0047049522399902
0.0047979354858398
0.0047011375427246
0.0047698020935059

Avg: 0.0048041582107544 sec
```


Regex:
```
> for ($i = 0; $i -lt 10; $i++) { php .\tempest }

0.00072097778320312
0.00059008598327637
0.00069689750671387
0.00060105323791504
0.00063681602478027
0.00059008598327637
0.00056695938110352
0.0006558895111084
0.00058507919311523
0.00063514709472656

Avg: 0.00062789916992187 sec
```

---

So, at the scale of thousandths of a second, the RegEx is faster. However, i lean toward @brendt's concern about reliable accuracy. But if the time difference is convincing enough, i'm open to switching to it.

My last thoughts on it: practically speaking, once caching is turned on, the time difference only matters once, right? Hell, discovery and caching the packages could be part of the composer post-install scripts, couldn't it? It never even has to occur for actual application requests/commands.

---

Anyway, how do we feel about this version? Any specific changes to make?

Comment (brendt) [reply src/Application/Kernel.php:159]: I say we go with the non-regex approach. Sorry @aidan-casey 🫣

Regarding cache: local development won't have it enabled, so we need it to be performant enough still. I think the current numbers look good enough

Comment (brendt) [reply src/Application/Kernel.php:159]: Vendor discovery can be cached locally btw, we're already planning on adding a discovery:cache command for that

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:181]: We shouldn't use env directly, instead use `AppConfig`

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:180]: I don't understand what this part of the code does? Why the check on environment? What's the difference between these lines?

```
$autoloads = [$composer['autoload']['psr-4'] ?? []];

// …

$autoloads[] = $composer['autoload']['psr-4'] ?? [];
```

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:202]: Does this need a separate function?

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:203]: I would write a more specific exception message. Probably create a new `KernelException`. 

> Could not locate installed.json in {$path}, try running `composer install`

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:184]: What's happening here exactly? Looks a bit weird to me that we first store everything in `$autoloads` and then merging it? Are you trying to flatten the array?

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:162]: Why wrapping in `realpath`? If this is necessary, we might need to add it to the `path` helper function so that it is consistent.

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:156]: rename to `$installedPath` for clarity

File: src/Application/Kernel.php
Comment (brendt) [inline src/Application/Kernel.php:164]: I would refactor this for clarity into two boolean variables:

```php
$requiresTempest = isset($package['require']['tempest/framework']);
$hasRootNamespaces = isset($package['autoload']['psr-4']) && is_array($package['autoload']['psr-4']);

if ($requiresTempest && $hasRootNamespaces) {
```

**Review by brendt — CHANGES_REQUESTED** [review]
I've added some remarks, could you take a look?

Good work btw, really appreciate it 👍

File: src/Application/Kernel.php
Comment (aidan-casey) [reply src/Application/Kernel.php:164]: I don't think we need the check for whether it's an array.

File: src/Application/Kernel.php
Comment (aidan-casey) [reply src/Application/Kernel.php:159]: Hey, I get it. You guys don't care about speed. 😢🙃

File: src/Application/Kernel.php
Comment (brendt) [reply src/Application/Kernel.php:164]: True, composer already validates that

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:180]: Oops. i left out part of the key name inside the `if`. Should look like:

```php
if ( /* not production */ ) {
    $autoloads[] = $composer['autoload-dev']['psr-4'] ?? [];
}
```

We always want the production `"autoload"` set of PSR-4 namespaces. But the original code was also including the framework's `"Tests\\Tempest\\": "tests/"` namespace, which is in `"autoload-dev"`.

I didn't want to include dev namespaces in a production environment, but to keep continuity with the original code, this adds the `"autoload-dev"` section for non-production.

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:162]: i wouldn't mind seeing it added to `path()` -- i get nervous with relative path segments, and when you try to read them in output or debugging it's easy to get confused. i prefer to have the resolved actual path whenever possible.

Your call: i can make the change to `path()` or remove `realpath()` here.

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:164]: Agreed on removing `is_array()` ... i'm too accustomed to code with unreliable data structures in it. i should _hope_ composer is more thoughtful than that. 😉 

As for separating the `isset()` ... i agree the variables describe the condition, and i'll make the change.

For what it's worth, i try to balance concise code with descriptive code. Separate variables, each calling `isset()`, felt like variable pollution. Whereas one call that checks both felt sufficient. Personal preference, i don't feel strongly enough about it to protest. 😃

File: src/Application/Kernel.php
Comment (brendt) [reply src/Application/Kernel.php:180]: I don't think we should take that into account here, we could manually add items to appconfig during test setup

Comment (jtheuerkauf) [reply src/Application/Kernel.php:180]: i'll cover this in your new question below.

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:184]: The original functionality that came from using `autoload_psr4.php` was _all_ the defined namespaces are added to discovery - including those in `"autoload-dev"`.

i wanted to preserve that original logic, but it doesn't make sense (to me) to include development level namespaces in production.

The merge keeps the final array unique, to prevent accidental duplication of a namespace key. Duplication might cause other problems, but this doesn't have to be one of them. 😉 

So, all of that said, i can alter it any way you wish. i can always include or always exclude the dev namespaces, which removes the environment check. i could also flip it so the dev set is added first, then production overwrites any accidental duplicate entry... Whatever you'd prefer.

```php
/*
 * Given this structure:
 * {
 *   "autoload": {
 *     "psr-4": {
 *       "Foo\\": "foo",
 *       "Bar\\": "bar"
 *     },
 *  },
 * "autoload-dev": {
 *   "psr-4": {
 *     "Tests\\": "tests",
 *     "Bar\\": "bar-tests" (hypothetical duplicate namespace!)
 *   }
 * }
 */


// Always include dev (prod overwrites dev)
$autoloads = array_merge($composer['autoload-dev']['psr-4'], $composer['autoload']['psr-4'])
/*
 * Result:
 * {
 *   "Tests\\": "tests",
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 */

// Always include dev (dev overwrites prod)
$autoloads = array_merge($composer['autoload']['psr-4'], $composer['autoload-dev']['psr-4'])
/*
 * Result:
 * {
 *   "Foo\\": "foo",
 *   "Tests\\": "tests",
 *   "Bar\\": "bar-tests"
 * }
 */

// Always exclude dev
$autoloads = $composer['autoload']['psr-4'];
/*
 * Result:
 * {
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 */

// Conditionally include dev (dev overwrites prod - current submission)
$autoloads = [ $composer['autoload']['psr-4'] ];
// if not-prod add dev section and merge the two sections to a single set of namespaces
/*
 * Result (prod):
 * {
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 * Result (not prod):
 * {
 *   "Foo\\": "foo",
 *   "Tests\\": "tests",
 *   "Bar\\": "bar-tests"
 * }
 */

// Conditionally include dev (prod overwrites dev)
if ( /* not-production */ ) {
    $autoloads = [ $composer['autoload-dev']['psr-4'] ];
}
$autoloads[] = $composer['autoload']['psr-4'];
$autoloads = array_merge(...$autoloads);
/*
 * Result (prod):
 * {
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 * Result (not prod):
 * {
 *   "Tests\\": "tests",
 *   "Foo\\": "foo",
 *   "Bar\\": "bar"
 * }
 */
```

Perhaps i'm really over-thinking it, but it seems like an important detail, so i want to be thorough...

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:202]: i took your suggestion for a `KernelException` class to throw when the target file isn't found, and put it here. It cleans this up a little _and_ the spot where `is_file()` was being checked. One less nesting level above, and more consistent flow here: always check for the file, then return the decoded file contents.

Does that make more sense?

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:162]: Hm. `realpath()` fails on glob patterns like `.../tempest-framework/src/Config/**.php` (coming from `Kernel::initConfig()`

That's unfortunate.

File: src/Application/Kernel.php
Comment (brendt) [reply src/Application/Kernel.php:184]: I think the best approach is to never look at autoload-dev for discovery. The only time we need it is for testing of the framework itself, which we can easily fix within `Tempest\Tests\TestCase::setUp()`

Don't worry about tests breaking if you omit the autoload-dev part, I'm happy to fix that once the rest is in place :) I need to tinker around with it a bit myself, so I can't give you a ready-solution, but don't worry about it :) 

Are you ok with omitting all autoload-dev detection?

File: src/Application/Kernel.php
Comment (brendt) [reply src/Application/Kernel.php:162]: Oh, that's unfortunate indeed. How about we just use `path` as is for now, and we tackle `realpath` problems in a separate PR or issue after this PR gets merged. I like it when PRs focus on one core problem and not diverge too much from the initial goal :)

**Comment by brendt** [comment]
@aidan-casey it might be a good idea, but first let's focus on get this merged in, then think about possible extra refactorings :)

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:162]: Roger that. It could be as simple as `return realpath($path) ?: $path;` but i don't feel great about "your input has been changed, maybe" results.

File: src/Application/Kernel.php
Comment (jtheuerkauf) [reply src/Application/Kernel.php:184]: i'm _completely_ OK with leaving the dev info out of it. 😄

**Comment by jtheuerkauf** [comment]
New commit pushed.

i agree there's a LOT of code dedicated to discovery in the Kernel right now, and maybe it belongs in a Bootloader or something (i've been tinkering in Spiral lately, so i see a lot of "register Bootloaders in Kernel" so the boot tasks (e.g. Discovery) can be compartmentalized. Probably not the only framework to do it, just what i've been practicing with lately.

**Comment by jtheuerkauf** [comment]
Thanks @brendt! Glad i could contribute something.

**Pull request merged**
